Add developer style guide to doxygen documentation (#1205)
aws/aws-iot-device-sdk-embedded-C: docs/doxygen/developer/style.txt

- Variables are always initialized.
- A separator is placed between different sections of a file. The current separator is:
@code{c}
/*-----------------------------------------------------------*/
@endcode
- All files must include the config file at the top of the file <b>before any other includes</b>.
- `static` functions must have a declaration at the top of the file and be implemented before any application-facing functions.
- The [GNU complexity](https://www.gnu.org/software/complexity/manual/complexity.html) of every function should be less than 9.
- Deviations from [MISRA C:2012](https://en.wikipedia.org/wiki/MISRA_C) coding guidelines should be documented as comments in the code.

@subsection guide_developer_styleguide_codingstyle_typeguidelines Type guidelines
@brief Guidelines for variable types.
- Although not a part of the C89 standard, only fixed-width integer types should be used. Exceptions for `bool` and types required by third-party APIs.
- The default integer in the libraries should be 32 bits wide, i.e. `int32_t` or `uint32_t`.
- Sizes and lengths should be represented with `size_t`.
- Libraries may define `bool` macros for use with C89 compilers.

> @subsection guide_developer_styleguide_codingstyle_example Example File

Author: Okay I like the way it looks too

 * declarations. Global variables are permitted, but should be avoided when
 * possible. */

/* Global variables are always initialized. */
static int globalVariable = 0;
static int globalArray[ LIBRARY_CONSTANT ] = { 0 };

/*-----------------------------------------------------------*/

/* Implementations of static functions follow the global variable declarations. */

static bool libraryStaticFunction( void * pArgument,
                                   size_t argumentLength )
{
    /* All local variables are declared at the top of the function. Variables are
     * always initialized. */
    size_t i = 0;
    int32_t localVariable = 0;
    int32_t * pLocalPointer = ( int32_t * ) pArgument;
    bool status = true;

     /* All functions make generous use of the logging library. */
    LogInfo( "Performing calculation..." );

    if( ( pArgument == NULL ) || ( argumentLength == 0 ) ) /* Note the parentheses and spacing in if statements */
    {
        LogError( "Bad parameters." );

        /* Local variable instead of a goto for error handling. */
        status = false;
    }
    else /* Note an else clause is used instead of goto for error handling */
    {

>         for( i = 0; i < argumentLength; i++ ) /* Note the spacing in the for loop. */

Author: I changed this to 
```c
int32_t localVariable = 0;
size_t i;

for( i = 0; i < argumentLength; i++ ) /* Note the spacing in the for loop. */
{
```

>         for( i = 0; i < argumentLength; i++ ) /* Note the spacing in the for loop. */

Reviewer: Wouldn't declaring `i` in the `for` loop would be a narrower scope than declaring it outside of the loop?

>         for( i = 0; i < argumentLength; i++ ) /* Note the spacing in the for loop. */

Author: Seems it's a violation of c89:
```
error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
```

>         for( i = 0; i < argumentLength; i++ ) /* Note the spacing in the for loop. */

Reviewer: Huh. TIL

bool Library_ApplicationFunction( void ) /* Functions with no arguments have void in their argument list. */
{
    LIBRARY_FUNCTION_MACRO( globalArray );

    return true;
}

/* Separator and newline at end of file */
/*-----------------------------------------------------------*/

@endcode

@section guide_developer_styleguide_naming Naming
@brief Naming convention used in this SDK.

The naming convention aims to differentiate this SDK's files, variables, and functions to avoid name collisions. In general:
- The first characters of all <i>publicly visible</i> names should identify the name as part of this SDK. <br>
  <b>Example:</b> For general-purpose libraries (such as MQTT), names start with <library-name>_ (i.e. MQTT_ for the MQTT API).
- Words in names should be ordered with the most general word first and the most specific word last. <br>
  <b>Example:</b> `core_mqtt_serializer.c` identifies a file as part of the general MQTT library. `MQTT_Connect` identifies a public-facing function of the general MQTT library.
- Names should avoid using abbreviations.

@subsection guide_developer_styleguide_naming_definedconstantsandenumvalues Defined constants and enum values

> @brief Naming convention for constants set using preprocessor @c #`define` and enum values.

Author: It does work. That was just from the original style guide.

@formattable{defined constants and enum values}
@formattableentry{<b>LibraryDescription</b>,Defined constants and enum values in application-facing library header files,`MQTTSuccess` (core_mqtt_lightweight.h)}
@formattableentry{<b>LIBRARY_DESCRIPTION</b>,Defined constants and enum values in demos and tests,`MQTT_EXAMPLE_TOPIC`}
@formattableentry{<b>DESCRIPTION</b>,Internal constants and enum values<br>No `AWS_` prefix for internal names in AWS-specific libraries,`ROOT_CA_CERT_PATH`}

@subsection guide_developer_styleguide_naming_files Files
@brief Naming convention for files.

@formattable{files}
@formattableentry{<b>library</b>`_`<b>description</b>`.extension`,General library file,`core_mqtt_serializer.c`}
@formattableentry{<b>library</b>`_internal.h`,Internal library header,`core_mqtt_internal.h`}
@formattableentry{<b>library</b>`_`<b>demo</b>`_description.c`,Library demo source,`mqtt_demo_plaintext.c`}

> @formattableentry{<b>library</b>`_description_`<b>utest</b>`.c`<br><b>library</b>`_description_`<b>system_test</b>`.c`,Library test source,`core_mqtt_utest.c`<br>`mqtt_system_test.c`}

Reviewer: The description said some the optional parts of this text were *italic*? If it's not, then we can remove that part.

- `core_mqtt_internal.h`: Declares the MQTT library's internal functions, structures, macros, etc.

<p>File names for demos should begin with `library_demo_`. 
Unit tests must start with `library_` and end with `_utest`. 
System tests must start with `library_` and end with `_system_test`. 
The names should then specify the library being demoed or tested. 
For example, the files names of the MQTT library's demos start with `mqtt_demo_`. 
Additionally, test file names should describe what tests are implemented in the file, 
such as `mqtt_demo_mutual_auth.c` for a file containing a demo that mutually 
authenticates over TLS with an MQTT broker.</p>

@subsection guide_developer_styleguide_naming_functions Functions (and function-like macros)
@brief Naming convention of functions and function-like macros.

@formattable{functions}
@formattableentry{<b>Library</b>`_`<b>Description</b>,Externally-visible library function,`MQTT_Publish`}
@formattableentry{<b>description</b>,`static` function (never uses `Aws` prefix),`sendPublish`<br>`eventCallback`}

> @formattableentry{`test_`<b>Library</b>`_`<b>accessedFunction_Descrption</b>,Test access function,`test_MQTT_ProcessLoop_Timer_Overflow`}

Reviewer: Description. Also, it looks like the description is optional? I don't think all test functions have one
```suggestion
@formattableentry{`test_`<b>Library</b>`_`<b>accessedFunction_</b><i>Description</i>,Test access function,`test_MQTT_ProcessLoop_Timer_Overflow`}
```

Author: Putting the italic before the underscore

Author: Actually, I'm removing the italic because it would be best to have a description for each test case

Externally visible (i.e. not `static`) functions are [UpperCamelCased](https://en.wikipedia.org/wiki/Camel_case). Function names should then specify their library name, followed by an underscore, followed by a brief description of what the function does. For example:
- `MQTT_SerializePublish`: This function is part of the public MQTT API. It serializes an MQTT PUBLISH packet into a buffer.

Functions not visible outside their source file (i.e. `static` functions) have names that are [lowerCamelCased](https://en.wikipedia.org/wiki/Camel_case). These function names do not contain the library name or `Aws`. For example:
- `receiveSingleIteration`: A `static` function in `core_mqtt.c`.

@subsection guide_developer_styleguide_naming_types Types
@brief Naming conventions of library `typedef` types.

@formattable{types}
@formattableentry{`Library`<b>Description</b>`_t`,General types in application-facing library header files,`MQTTStatus_t` (core_mqtt.h)}

> @formattableentry{`Library`<b>Function</b>`Info_t`,Application-facing parameter structure,`MQTTPublishInfo_t`<br>(Parameter structure to `MQTT_Publish`)}

Reviewer: We also have `MQTTPacketInfo_t` and `MQTTDeserializedInfo_t`, which don't really fit with this. I'd probably just remove the second line

Types intended for internal library use defined in a header file are [lowerCamelCased](https://en.wikipedia.org/wiki/Camel_case). The names must start with the library name and end with `_t`. Internal types defined in a library source file must end with `_t` and not include the library name.

`struct` typedefs should always be named along with the `typedef`. The struct name should be identical to the `typedef` name, but without the `_t` suffix. For example:
@code{c}
typedef struct LibraryStruct
{
    int32_t member;
} LibraryStruct_t;

typedef struct libraryInternalStruct
{
    int32_t member;
} libraryInternalStruct_t;
@endcode

A `struct` may contain `union` members, but its usage must be documented as it violates MISRA rule 19.2.

@subsection guide_developer_styleguide_naming_variables Variables
@brief Naming conventions of variables.

@formattable{variables}
@formattableentry{<b>variableDescription</b>,General local variable,`startTime`}

> @formattableentry{`p`<b>VariableDescription</b>,Local variable pointers and arrays (including strings),`pSubscriptionList`}

Reviewer: The line below says it also applies to globals
```suggestion
@formattableentry{`p`<b>VariableDescription</b>,Variable pointers and arrays (including strings),`pSubscriptionList`}
```

/**
@page guide_developer_styleguide Style Guide
@brief Guide for the coding style used in this SDK.

The goal of this style guide is to encourage a readable and consistent coding style across the entire SDK.

@section guide_developer_styleguide_codingstyle Coding Style
@brief The coding style used in this SDK.

The coding style aims to produce code that is readable and easy to debug. An example is provided in @ref guide_developer_styleguide_codingstyle_example.

@subsection guide_developer_styleguide_codingstyle_generalguidelines General guidelines
@brief General guidelines for library style.

> - All new libraries (being developed in the development branch) should only use [ANSI C](https://en.wikipedia.org/wiki/ANSI_C#C89) features and standard library functions. (All libraries in the v4_beta branch used the C99 standard, so any changes to the branch can continue using C99).

Reviewer: We can take out the part about the development branch. Do we still need to mention v4_beta?

Author: Removed v4_beta part

> - Libraries should log extensively.

Reviewer: Didn't we decide on a caveat to this? Something about not needing to log for pure functions? @dan4thewin

Author: Updated to say: ```Libraries should log extensively except in the case of pure functions.```

- Code should be well-commented.

> - Only <code>`/`*</code> and <code>*`/`</code> should be used to start and end comments.

Reviewer: Do we need `` if we already have the `<code>` tags?

Author: `<code>` can be taken out, but it looks better with it. The `` is so that doxygen doesn't recognize it as a comment or closing tag. Let me know if you know another way around it.

- All comments end with a period.
- Use of `goto` is discouraged.
- Only spaces should be used for indenting. A single indent is 4 spaces. No tab characters should be used.
- A parenthesis is usually followed by a space (see @ref guide_developer_styleguide_codingstyle_example).
- Lines of code should be less than 80 characters long, although longer lines are permitted.

> - Local variables should be declared at the top of a block in the narrowest scope possible, at the programmer's discretion.

Reviewer: Maybe we should mention that reducing complexity could be a valid case for not declaring them in the narrowest scope

- All global variables should be declared at the top of a file.
- Variables are always initialized.
- A separator is placed between different sections of a file. The current separator is:
@code{c}
/*-----------------------------------------------------------*/
@endcode
- All files must include the config file at the top of the file <b>before any other includes</b>.
- `static` functions must have a declaration at the top of the file and be implemented before any application-facing functions.
- The [GNU complexity](https://www.gnu.org/software/complexity/manual/complexity.html) of every function should be less than 9.
- Deviations from [MISRA C:2012](https://en.wikipedia.org/wiki/MISRA_C) coding guidelines should be documented as comments in the code.

@subsection guide_developer_styleguide_codingstyle_typeguidelines Type guidelines
@brief Guidelines for variable types.
- Only fixed-width integer types should be used. Exceptions for `bool` and types required by third-party APIs.
- The default integer in the libraries should be 32 bits wide, i.e. `int32_t` or `uint32_t`.

> - Sizes and lengths should be represented with `size_t`, and Boolean variables with `bool`.

Reviewer: We might want to give more clarification on this. Neither fixed width integers nor bools are a part of C89. Something like a line saying "Although not a part of the C89 standard, libraries may define `bool` macros may be defined for use with C89 compilers."

Author: - Although not a part of the C89 standard, only fixed-width integer types should be used. Exceptions for `bool` and types required by third-party APIs.
- The default integer in the libraries should be 32 bits wide, i.e. `int32_t` or `uint32_t`.
- Sizes and lengths should be represented with `size_t`.
- Libraries may define `bool` macros for use with C89 compilers.

/* Defined constants follow the included headers. */

/* When possible, parentheses () should be placed around constant values and a type
 * should be specified. */
#define LIBRARY_CONSTANT ( ( int32_t ) 10 )

#define LIBRARY_FUNCTION_MACRO( argument ) \ /* Line continuators are right-aligned. */
    {                                      \ /* Function-like macros are surrounded by curly braces {}. */
        macro_body( ( argument ) );        \ /* Parentheses surround macro arguments for MISRA 20.7 */
    }

/*-----------------------------------------------------------*/

/* Library typedefs follow the defined constants. */

/* Forward declarations are used only when necessary. They are placed before all
 * other typedefs. */

typedef int32_t type_t;

/**
   @brief Structs are named along with the typedef. 
*/
typedef struct structType 
{
    int32_t member;

    /* As usage of unions violates MISRA rule 19.2, we must document our usage
     * and justification. Unions are used here to reduce the size of this struct. */
    union /* Anonymous structs/unions are permitted only inside of other structs. */
    {
        int32_t a;

>         int32_t b;

Reviewer: When would you ever have a union of two `int32_t`s...?

Author: Just copied from the wiki. I don't think we even allow unions, so I'm removing this.

    };

   /* Although variable length arrays (a C99 feature) violate MISRA rule 18.7,
    * they are permitted provided documentation is supplied in comments. */
   int32_t variableLengthMember[];
} structType_t;

/*-----------------------------------------------------------*/

/* Declarations of static and extern functions follow the typedefs. */

static bool libraryStaticFunction( void * pArgument,
                                   size_t argumentLength );

/* External function declarations should be used sparingly (using an internal
 * header file to declare functions is preferred). */
extern int32_t Library_ExternalFunction( void * pArgument );

/*-----------------------------------------------------------*/

/* Declarations of global variables follow the static and extern function
 * declarations. Global variables are permitted, but should be avoided when
 * possible. */

/* Global variables are always initialized. */
static int globalVariable = 0;

> static int globalArray[ _LIBRARY_CONSTANT ] = { 0 };

Reviewer: We're no longer allowing identifiers to begin with underscores, I think.

@formattable{files}
@formattableentry{<b>library</b>`_`<b>description</b>`.extension`,General library file,`core_mqtt_serializer.c`}
@formattableentry{<b>library</b>`_internal.h`,Internal library header,`mqtt_internal.h`}
@formattableentry{<b>library</b>`_`<b>demo</b>`_description.c`,Library demo source,`mqtt_demo_plaintext.c`}
@formattableentry{<b>library</b>`_description_`<b>utest</b>`.c`<br><b>library</b>`_description_`<b>system_test</b>`.c`,Library test source,`core_mqtt_utest.c`<br>`mqtt_system_utest.c`}

File names contain only lowercase letters and underscores. All file names should be named according to their purpose. For example:
- `mqtt_api.c`: A file in the MQTT library that implements the MQTT API functions.
- `demo_mqtt.c`: A file in the Demos for the MQTT library.
- `mqtt_demo_plaintext.c`: A file in the demos for the MQTT library that communicates over a plaintext channel.
- `core_mqtt_utest.c`: A file in the Tests for the MQTT library. Since the tests currently only run on POSIX systems, test file names do not use the `_posix` suffix.

Library file names should use one or two words to describe the functions implemented in that file. For example:

> - `mqtt_serialize.c`: Implements the MQTT library's packet serialization and deserialization functions.

Reviewer: The file's called `core_mqtt_serializer.c` though

@brief Naming conventions of library `typedef` types.

@formattable{types}
@formattableentry{`Library`<b>Description</b>`_t`,General types in application-facing library header files,`MQTTStatus_t` (core_mqtt.h)}
@formattableentry{`Library`<b>Function</b>`Info_t`,Application-facing parameter structure,`MQTTPublishInfo_t`<br>(Parameter structure to `MQTT_Publish`)}
@formattableentry{<b>libraryDescription</b>`_t`,Type in an `internal` header,`mqttOperation_t`<br>`shadowOperation_t`}
@formattableentry{<b>description</b>`_t`,Internal type in source file,`topicMatchParams_t`}

Types intended for use in applications are [UpperCamelCased](https://en.wikipedia.org/wiki/Camel_case). The names end with `_t`. Parameter structures should indicate their associated function: for example, `MQTTPublishInfo_t` is passed as a parameter to `MQTT_Publish`.

Types intended for internal library use defined in a header file are [lowerCamelCased](https://en.wikipedia.org/wiki/Camel_case). The names must start with the library name and end with `_t`. Internal types defined in a library source file must end with `_t` and not include the library name.

> `struct` typedefs should always be named along with the `typedef`. The struct name should be identical to the `typedef` name, but without the `_t` at the end. For example:

Reviewer: ```suggestion
`struct` typedefs should always be named along with the `typedef`. The struct name should be identical to the `typedef` name, but without the `_t` suffix. For example:
```

@code{c}
typedef struct LibraryStruct
{
    int32_t member;
} LibraryStruct_t;

typedef struct libraryInternalStruct
{
    int32_t member;
} libraryInternalStruct_t;
@endcode

A `struct` may <b>NOT</b> contain `union` members.

@subsection guide_developer_styleguide_naming_variables Variables
@brief Naming conventions of variables.

@formattable{variables}
@formattableentry{<b>variableDescription</b>,General local variable,`startTime`}
@formattableentry{`p`<b>VariableDescription</b>,Local variable pointers and arrays (including strings),`pSubscriptionList`}
@formattableentry{`_`<b>variableDescription</b>,General global variable,`lastSubscribePacketIdentifier`}}

Local variable names are [lowerCamelCased](https://en.wikipedia.org/wiki/Camel_case) and consist only of a description of the variable. Names like `i` or `j` are acceptable for loop counters, but all other variables should have a descriptive name.

> Global variable names always start with a `_`. Global variables that are `static` consist of only the description in [lowerCamelCase](https://en.wikipedia.org/wiki/Camel_case). Global variables that are not static consist of the library name <b>and</b> the description in [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case).

Reviewer: Not anymore

/* Standard includes. */
#include <assert.h>
#include <stddef.h>
#include <string.h>

> /* bool is defined in only C99+. */

Reviewer: I think a better description would be
```suggestion
/* This file defines `bool` for compatibility with C89 compilers. */
```
